src/aio.rs

```diff
@@ -0,0 +1,325 @@
+        let rc = unsafe {
+            // It's safe because parameters are valid and we have checked the result.
+            libc::syscall(
+                libc::SYS_io_cancel,
+                self.0,
+                iocb as *const IoControlBlock,
+                result as *mut IoEvent,
+            ) as c_int
+        };
+        if rc < 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Read asynchronous I/O events from the completion queue.
+    ///
+    /// Refer to Linux [`io_getevents`](http://man7.org/linux/man-pages/man2/io_getevents.2.html).
+    ///
+    /// # Arguments
+    /// * `min_nr`: read at least min_nr events.
+    /// * `events`: array to receive the io operation results.
+    /// * `timeout`: optional amount of time to wait for events.
+    pub fn get_events(
+        &self,
+        min_nr: c_long,
+        events: &mut [IoEvent],
+        timeout: Option<&mut libc::timespec>,
+    ) -> Result<usize> {
+        let to = match timeout {
+            Some(val) => val as *mut libc::timespec,
+            None => null_mut() as *mut libc::timespec,
+        };
+
+        // It's safe because parameters are valid and we have checked the result.
+        let rc = unsafe {
+            libc::syscall(
+                libc::SYS_io_getevents,
+                self.0,
+                min_nr,
+                events.len() as c_long,
+                events.as_mut_ptr(),
+                to,
+            ) as c_int
+        };
+        if rc < 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(rc as usize)
+        }
+    }
+}
+
+impl Drop for IoContext {
+    fn drop(&mut self) {
+        if self.0 != 0 {
+            // It's safe because the context is created by us.
+            let _ = unsafe { libc::syscall(libc::SYS_io_destroy, self.0) as c_int };
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::fs::File;
+    use std::os::unix::io::AsRawFd;
+
+    #[test]
+    fn new_context() {
+        let _ = IoContext::new(0).unwrap_err();
+    }
+
+    #[test]
+    fn cancel_request() {
+        let file = File::open("/dev/zero").unwrap();
+
+        let ctx = IoContext::new(128).unwrap();
+        let mut buf: [u8; 16384] = unsafe { std::mem::uninitialized() };
+        let iocbs = [&mut IoControlBlock {
+            aio_fildes: file.as_raw_fd() as u32,
+            aio_lio_opcode: IOCB_CMD_PREAD as u16,
+            aio_buf: buf.as_mut_ptr() as u64,
+            aio_nbytes: buf.len() as u64,
+            ..Default::default()
+        }];
+
+        let mut rc = ctx.submit(&iocbs).unwrap();
+        assert_eq!(rc, 1);
+
+        let mut result = Default::default();
+        let err = ctx
+            .cancel(iocbs[0], &mut result)
+            .unwrap_err()
+            .raw_os_error()
+            .unwrap();
+        assert_eq!(err, libc::EINVAL);
+
```

Comment:
So here we tested a failure cancel() operation right? So do we need check `result`? Meanwhile I am thinking if we need test a successful one?

Reply:
If you read linux kernel implementation, you will find that aio::cancel() won't success at all:(
Aio is a wired subsystem, and seems no plan to fix it at all.

Reply:
Thanks for the explanation. So we might comment a few words in cancel() implementation, WDYT? And others seems good to me, besides the comment of copying some Examples. :)

Reply:
done:)